Sync/Synchronizers/Bookmarks/BookmarksSynchronizer.swift

```diff
@@ -120,18 +120,66 @@ open class BufferingBookmarksSynchronizer: TimestampedSingleCollectionSynchroniz
 
         if !AppConstants.shouldMergeBookmarks {
             run = doMirror >>== { result in
-                // Just validate to report statistics.
-                if case .completed = result {
-                    log.debug("Validating completed buffer download.")
-                    return buffer.validate().bind { validationResult in
-                        if let invalidError = validationResult.failureValue as? BufferInvalidError {
-                            self.statsSession.validationStats = self.validationStatsFrom(error: invalidError)
+                guard case .completed = result else {
+                    return deferMaybe(result)
+                }
+                return storage.getLocalBookmarksAdditions() >>== { newBookmarks -> Success in
+                    guard newBookmarks.count > 0 else {
+                        return succeed()
+                    }
+                    // This assumes that "Mobile Bookmarks" exists in the buffer, which will be false if no computer ever connected to that sync account
+                    return buffer.getBufferItemAndChildrenWithGUID(BookmarkRoots.MobileFolderGUID) >>== { curMobileRoot in
+                        let newChildren = newBookmarks.values.map { $0.guid }
+                        var allChildren = curMobileRoot.children ?? []
+                        // We do not use a set in order to keep the children's order
+                        for child in newChildren {
+                            if !allChildren.contains(child) {
```

Comment:
You're fetching and materializing potentially hundreds of records so that you can do a containment check. That containment check should only fail if we were interrupted on an upload, and in any case the check is wrong — you need to make sure the GUID doesn't exist anywhere on the server!

So let's use SQL for this: the set of records you want to upload is the set of children of the mobile root in the local table, minus the set of GUIDs in the buffer, minus the set of records in the mirror (which should always be empty, but let's be sure). Call that set `newRecordGUIDs`.

If there's no mobile root in the buffer, create one with `children = newRecordGUIDs`. Otherwise, take the existing root and append `newRecordGUIDs`.

Upload all of the local records in `newRecordGUIDs`, plus the amended mobile root. You don't need to fetch the existing children as bookmarks at all.